### PR TITLE
ci(golangci-lint): add unparam linter

### PR DESCRIPTION
**Pull request by acud**
## Motivation

Adds the `unparam` linter.

## TODO

- [ ] Explain motivation or link existing issue(s)
- [ ] Test changes and document test plan
- [ ] Update documentation as needed
- [ ] Update [changelog](../CHANGELOG.md) as needed

### CODE REVIEW

File: atxsdata/warmup_test.go [hunk 1]
```diff
@@ -17,11 +17,12 @@ import (
 	"github.com/spacemeshos/go-spacemesh/sql/mocks"
 )
 
+var nonce = types.VRFPostIndex(1)
```
Comment (poszu) [inline atxsdata/warmup_test.go:20]: How about inlining `types.VRFPostIndex(1)` in `gatx()` instead of creating a global?

File: beacon/handlers_test.go [hunk 2]
```diff
@@ -17,14 +17,18 @@ import (
 	"github.com/spacemeshos/go-spacemesh/signing"
 )
 
-const epochWeight = uint64(100)
+const (
+	epochWeight = uint64(100)
+	// FIXME: pulled out to satisfy the linter but needs to be
+	// actually modified in tests to run different values.
+	epoch = types.EpochID(10)
+	round = types.RoundID(5)
```
Comment (poszu) [inline beacon/handlers_test.go:25]: Then perhaps add a `//nolint` with an explanation to the function definition instead?

File: genvm/vm_test.go [hunk 3]
```diff
@@ -465,11 +465,13 @@ func (t *tester) estimateSpendGas(principal, to, amount int, nonce core.Nonce) i
 		int(core.TxDataGas(len(tx)))
 }
 
-func (t *tester) estimateDrainGas(principal, vault, to, amount int, nonce core.Nonce) int {
+const vaultVal = 20
+
+func (t *tester) estimateDrainGas(principal, to, amount int, nonce core.Nonce) int {
```
Comment (poszu) [inline genvm/vm_test.go:470]: Is this the right way to fix it? Perhaps the tests should use different values?

File: hare3/protocol_test.go [hunk 4]
```diff
@@ -50,9 +50,13 @@ func (t *tinput) round(r Round) *tinput {
 	return t
 }
 
-func (t *tinput) iter(i uint8) *tinput {
+// FIXME: made to satisfy the linter but actually needs unit test
+// changes to test different cases.
+const iter1 = uint8(1)
```
Comment (poszu) [inline hare3/protocol_test.go:55]: How about using //nolint then?

File: hare3/hare_test.go [hunk 5]
```diff
@@ -753,7 +753,9 @@ func TestHandler(t *testing.T) {
 	})
 }
 
-func gatx(id types.ATXID, epoch types.EpochID, smesher types.NodeID, base, height uint64) types.ActivationTx {
+const epoch = types.EpochID(1)
```
Comment (poszu) [inline hare3/hare_test.go:756]: I'm not sure about this linter if enabling it ends up using globals. The flow of information is broken now and the code is more difficult to follow.

File: api/grpcserver/post_service_test.go [hunk 6]
```diff
@@ -74,7 +73,6 @@ func launchPostSupervisor(
 func launchPostSupervisorTLS(
```
Comment (fasmat) [inline api/grpcserver/post_service_test.go:73]: This and `launchPostSupervisor` are basically identical, except that `launchPostSupervisor` doesn't yet use `sql.InMemoryTest()` and this function has the wrong name for the logger of the post manager.

I think `launchPostSupervisorTLS` can be removed and in the one place where it is called be replaced with `launchPostSupervisor`

File: atxsdata/warmup_test.go (hunk 1 above)
Comment (fasmat) [reply atxsdata/warmup_test.go:20]: Also ideally it should not be the same value every time 😅

File: fetch/cache_test.go [hunk 7]
```diff
@@ -15,11 +15,12 @@ import (
...
 
 // getCachedEntry is a thread-safe cache get helper.
-func getCachedEntry(cache *HashPeersCache, hash types.Hash32) (HashPeers, bool) {
+func getCachedEntry(cache *HashPeersCache, hash types.Hash32) HashPeers {
 	cache.mu.Lock()
 	defer cache.mu.Unlock()
 
-	return cache.get(hash)
+	hp, _ := cache.get(hash)
+	return hp
```
Comment (fasmat) [inline fetch/cache_test.go:23]: I feel like either here or where this helper is called an assertion for the boolean is missing (I believe it should always be true in the current tests).

File: genvm/vm_test.go [hunk 8]
```diff
@@ -1459,10 +1461,12 @@ func runTestCases(t *testing.T, tcs []templateTestCase, genTester func(t *testin
 	}
 }
 
-func testWallet(t *testing.T, defaultGasPrice int, template core.Address, genTester func(t *testing.T) *tester) {
+const defaultGasPrice = 1
```
Comment (fasmat) [inline genvm/vm_test.go:1464]: The `defaultGasPrice` was 20 before, why the change to 1? Also should this maybe be just different values for different tests?

File: tortoise/state.go [hunk 9]
```diff
@@ -434,7 +434,7 @@ type headerWithSign struct {
 	sign   sign
 }
 
-func decodeVotes(evicted, blid types.LayerID, base *ballotInfo, exceptions types.Votes) (votes, types.LayerID, error) {
+func decodeVotes(blid types.LayerID, base *ballotInfo, exceptions types.Votes) (votes, types.LayerID, error) {
```
Comment (fasmat) [inline tortoise/state.go:437]: That seems suspicious, are we sure that `evicted` is not needed and just wasn't forgotten when implementing `decodeVotes`?

File: systest/tests/smeshing_test.go [hunk 10]
```diff
@@ -381,7 +381,12 @@ func genKeys(tb testing.TB, n int) (pks []ed25519.PrivateKey, pubs []ed25519.Pub
 	return pks, pubs
 }
 
-func prepareVesting(tb testing.TB, keys, start, end, initial, total int) vestingAcc {
+const (
+	start = 8
+	end   = 20
+)
```
Comment (fasmat) [inline systest/tests/smeshing_test.go:387]: I feel like instead of making these constant the test above should just use different start and end layers for vesting. Wdyt?

File: miner/active_set_generator_test.go [hunk 11]
```diff
@@ -58,8 +58,8 @@ type test struct {
 	expectErr string
 }
 
-func unixPtr(sec, nsec int64) *time.Time {
-	t := time.Unix(sec, nsec)
+func unixPtr(sec int64) *time.Time {
+	t := time.Unix(sec, 0)
 	return &t
 }
```
Comment (fasmat) [inline miner/active_set_generator_test.go:64]: This function isn't needed in my opinion. `epochStart` can just be a value instead of a pointer and instead of checking it for `nil` one can call `epochStart.IsZero()` on it to see if it was set or not.

Wdyt?

**Review by fasmat — CHANGES_REQUESTED** [review]

File: hare3/hare_test.go (hunk 5 above)
Comment (fasmat) [reply hare3/hare_test.go:756]: Also I believe `epoch` should in at least one test be a different value from the others - or are all these tests seriously only interested in the very first epoch after genesis? I.e. the one where nothing happens except ATXs are published?

File: hare3/protocol_test.go [hunk 12]
```diff
@@ -182,8 +186,10 @@ type setup struct {
 	proposals []types.ProposalID
 }
 
-func (s *setup) thresh(v uint16) *setup {
-	s.threshold = v
+const threshold = 10
```
Comment (fasmat) [inline hare3/protocol_test.go:189]: Same as above, probably `// nolint` is the way to go and fix it at some later point.

File: hare4/hare_test.go [hunk 13]
```diff
@@ -928,7 +928,9 @@ func TestHandler(t *testing.T) {
 	})
 }
 
-func gatx(id types.ATXID, epoch types.EpochID, smesher types.NodeID, base, height uint64) types.ActivationTx {
+const epoch = types.EpochID(1)
```
Comment (fasmat) [inline hare4/hare_test.go:931]: Same as above, epoch 1 is different from all other epochs, we shouldn't only test for that specific. We should have at least a single test where epoch is a different value.

File: hare4/protocol_test.go [hunk 14]
```diff
@@ -143,9 +143,13 @@ func (t *toutput) round(r Round) *toutput {
 	return t
 }
 
-func (t *toutput) iter(i uint8) *toutput {
+// FIXME: made to satisfy the linter but actually needs unit test
+// changes to test different cases.
+const iter1 = uint8(1)
```
Comment (fasmat) [inline hare4/protocol_test.go:148]: Just like in the other places, I believe we should have variable values for this in different tests and if we don't change it within this PR, then `// nolint` for now.

File: hare4/protocol_test.go [hunk 15]
```diff
@@ -182,8 +186,10 @@ type setup struct {
 	proposals []types.ProposalID
 }
 
-func (s *setup) thresh(v uint16) *setup {
-	s.threshold = v
+const threshold = 10
```
Comment (fasmat) [inline hare4/protocol_test.go:189]: Same as above

File: miner/proposal_builder_test.go [hunk 16]
```diff
@@ -145,9 +147,10 @@ func expectMeshHash(hash types.Hash32) expectOpt {
 	}
 }
 
+const epochExpectCounter = types.EpochID(3)
```
Comment (fasmat) [inline miner/proposal_builder_test.go:150]: This should probably always be `epoch+1` and `epoch` should again vary between tests (or `// nolint`)

**Comment by acud** [comment]
Thanks for the review guys, but the intent here was to enable the linter, however, I think that addressing all this will make this PR huge and unreasonable to review. We should probably fix the underlying issues and then enable the linter. Thanks

File: proposals/eligibility_validator_test.go [hunk 17]
```diff
@@ -83,8 +84,10 @@ func geligibilities(js ...uint32) (rst []types.VotingEligibility) {
 	return rst
 }
 
+const j = uint32(1)
```
Comment (fasmat) [inline proposals/eligibility_validator_test.go:87]: This also feels like it should vary between tests

File: sql/transactions/transactions_test.go [hunk 18]
```diff
@@ -17,11 +17,13 @@ import (
 	"github.com/spacemeshos/go-spacemesh/sql/transactions"
 )
 
+const amount = uint64(191)
```
Comment (fasmat) [inline sql/transactions/transactions_test.go:20]: I think this should vary between tests.

File: tortoise/fixture_test.go [hunk 19]
```diff
@@ -125,12 +127,14 @@ type bopt struct {
 	opts []ballotOpt
 }
 
-func (b *bopt) beacon(value string) *bopt {
+const beaconVal = "a"
```
Comment (fasmat) [inline tortoise/fixture_test.go:130]: This can probably be inlined

File: tortoise/fixture_test.go [hunk 20]
```diff
@@ -77,9 +77,11 @@ func (s *smesher) rawatx(id types.ATXID, publish types.EpochID, opts ...*aopt) *
 	return val
 }
 
+const publish = 1
```
Comment (fasmat) [inline tortoise/fixture_test.go:80]: As stated before, epoch 1 is special different from all other epochs, we should have at least one test were we have a different publish epoch.
